### PR TITLE
Replace elasticsearch in favor of opensearch as fts engine

### PR DESCRIPTION
**Pull request by Ark74**
Close https://github.com/nextcloud/vm/issues/1794

To try this commit, please install a base VM without FTS.
Then manually get and run this PR [fulltextsearch.sh](https://raw.githubusercontent.com/Ark74/vm/add_opensearch_fts_engine/apps/fulltextsearch.sh) app

That should get the app and engine working.
Some values are hard coded in order to make the PR work.

Regards.

### CODE REVIEW

**Comment by Ark74** [comment]
Closing as latest "fix" introduced a regression.
Will report soon.

**Comment by Ark74** [comment]
Ready for review :+1:

File: apps/fulltextsearch.sh
Comment (szaimen) [inline apps/fulltextsearch.sh:11]: todo: needs to be reverted

File: apps/fulltextsearch.sh
Comment (szaimen) [inline apps/fulltextsearch.sh:92]: Would it be possible to do it without requiring docker compose or is that not feasible anymore?

File: apps/fulltextsearch.sh
Comment (szaimen) [inline apps/fulltextsearch.sh:224]:
```suggestion
chmod 740 -R  $OPNSDIR
```

File: apps/fulltextsearch.sh
Comment (szaimen) [inline apps/fulltextsearch.sh:231]: do we really need a restart here?

File: lib.sh
Comment (szaimen) [inline lib.sh:183]: I think we can get rid of this line finally

File: apps/fulltextsearch.sh
Comment (szaimen) [inline apps/fulltextsearch.sh:216]: todo?

**Review by szaimen — COMMENTED** [review]
Thanks for doing this! looks like a pretty immense change.
Just had a quick look but didn't test.

File: apps/fulltextsearch.sh
Comment (szaimen) [inline apps/fulltextsearch.sh:50]: we should probably leave this here for cleaning up any potential old installation

File: apps/fulltextsearch.sh
Comment (szaimen) [inline apps/fulltextsearch.sh:95]: we should probably remove that file on old installations if they reinstall with this script?

File: apps/fulltextsearch.sh
Comment (szaimen) [reply apps/fulltextsearch.sh:50]: we should probably also remove the esdata volume in that case

File: apps/fulltextsearch.sh
Comment (szaimen) [inline apps/fulltextsearch.sh:52]: what about also resetting the fts_os-data volume?

File: apps/fulltextsearch.sh
Comment (Ark74) [reply apps/fulltextsearch.sh:50]: Oh! i didn't realized I've removed that. I'll restore it.
IIRC docker system prune -a should remove all images and volumes, right?

Comment (szaimen) [reply apps/fulltextsearch.sh:50]: > system prune -a should remove all images

it removes all dangling images and networks, afaik. volumes need to be purge with docker volume prune but I am not sure if thats the best way to do it. deleting this volume specifically is probably better...

File: apps/fulltextsearch.sh
Comment (Ark74) [reply apps/fulltextsearch.sh:216]: yeah the content is below.

File: apps/fulltextsearch.sh
Comment (szaimen) [reply apps/fulltextsearch.sh:92]: what I meant was if it is not possible to simply use a docker run command for the container instead of relying on docker-compose?

File: apps/fulltextsearch.sh
Comment (Ark74) [reply apps/fulltextsearch.sh:231]: not really, i just replace the old variable.

Comment (szaimen) [reply apps/fulltextsearch.sh:231]: okay, would be cool to get rid of that then :)

File: apps/fulltextsearch.sh
Comment (szaimen) [reply apps/fulltextsearch.sh:216]: okay

File: apps/fulltextsearch.sh
Comment (szaimen) [reply apps/fulltextsearch.sh:50]: yes but as I said i think it is better to just remove the specific volume here...

File: apps/fulltextsearch.sh
Comment (Ark74) [reply apps/fulltextsearch.sh:92]: we are not running a vanilla image, there are some addons installed, that will require a third party to host and maintain that new image, with docker-compose there is no need to do that.

File: apps/fulltextsearch.sh
Comment (Ark74) [reply apps/fulltextsearch.sh:224]: That would require testing

File: apps/fulltextsearch.sh
Comment (szaimen) [reply apps/fulltextsearch.sh:92]: okay, that is a good reason. Thanks for making that clear!

File: apps/fulltextsearch.sh
Comment (Ark74) [reply apps/fulltextsearch.sh:224]: this is  a no go.

File: lib.sh
Comment (Ark74) [reply lib.sh]: done

File: apps/fulltextsearch.sh
Comment (Ark74) [reply apps/fulltextsearch.sh:50]: fixed by 61bd632

File: apps/fulltextsearch.sh
Comment (Ark74) [reply apps/fulltextsearch.sh:95]: solved by 74616ea

File: apps/fulltextsearch.sh
Comment (Ark74) [reply apps/fulltextsearch.sh:53]: solved by 61bd632f40c6bdd2a20cec8075104f0aa1e27b2e

File: lib.sh
Comment (Ark74) [inline lib.sh:181]: TODO: replace when merged

File: apps/fulltextsearch.sh
Comment (szaimen) [inline apps/fulltextsearch.sh:55]: is docker_prune_this really able to prune volumes?

File: apps/fulltextsearch.sh
Comment (szaimen) [inline apps/fulltextsearch.sh:32]:
```suggestion
if ! does_this_docker_exist "$nc_fts" || ! does_this_docker_exist "$opens_fts" || ! is_app_installed fulltextsearch
```

**Review by szaimen — COMMENTED** [review]
Thanks, looks already much better!

File: apps/fulltextsearch.sh
Comment (szaimen) [inline apps/fulltextsearch.sh:32]:
```suggestion
if ! does_this_docker_exist "$nc_fts" && ! does_this_docker_exist "$opens_fts" && ! is_app_installed fulltextsearch
```

Comment (szaimen) [reply apps/fulltextsearch.sh:32]: Explanation: we only want to show the install_popup if neither of those containers/app was found. If it was, we want to show the option for reinstallation or uninstallation.

File: apps/fulltextsearch.sh
Comment (Ark74) [reply apps/fulltextsearch.sh:55]: actually it only stop them, then it prunes everything is not working.
https://github.com/nextcloud/vm/blob/74616ea74ea9ee3fa34ec7875e71d02f94b0805d/lib.sh#L1636-L1647

Comment (szaimen) [reply apps/fulltextsearch.sh:55]: ah okay, fine by me then

File: apps/fulltextsearch.sh
Comment (Ark74) [reply apps/fulltextsearch.sh:32]: yeah, ok.

**Comment by Ark74** [comment]
Please check the latest changes and let me know if it works for you.
Regards

**Comment by szaimen** [comment]
Thanks! lets wait for a review from @enoch85

**Comment by enoch85** [comment]
Hey guys, I will have a look at this next week. I'm sick in Covid here (second time) and need to rest. A little dead right now after been working anyway the whole day.

**Comment by szaimen** [comment]
> Hey guys, I will have a look at this next week. I'm sick in Covid here (second time) and need to rest. A little dead right now after been working anyway the whole day.

Rest and get well soon!

**Comment by Ark74** [comment]
Hello, sorry to read that.
I was like that last week, same 2nd time.
This week I'm dealing only the small symptoms.
Get well soon.

**Comment by szaimen** [comment]
- [ ] TODO: update the image during in the nextcloud_update script

**Comment by enoch85** [comment]
OK @Ark74, you wanted me to implement something in lib?

File: apps/fulltextsearch.sh
Comment (enoch85) [inline apps/fulltextsearch.sh:32]: Sometimes the docker exists, but not the app. Sometimes the app exists, but not the docker. It should check for all ocurrances seperatly and ask to reinstall if any of them are true.

File: static/opensearch_certs.sh
Comment (enoch85) [inline static/opensearch_certs.sh:1]: Can you please move this to /apps instead?

File: lib.sh
Comment (enoch85) [inline lib.sh:180]: change to: `run_script APPS opensearch_certs.sh`

File: lib.sh
Comment (enoch85) [reply lib.sh:181]: replace it now, we can fix bugs when merged. I trust that you tested it already.

**Review by enoch85 — CHANGES_REQUESTED** [review]

**Comment by Ark74** [comment]
@enoch85 it's already there, 
https://github.com/nextcloud/vm/blob/a7fa9a2a11bf8585f463a32c739f2032e8e12d6c/lib.sh#L178-185

just make sure that 
* https://github.com/nextcloud/vm/blob/a7fa9a2a11bf8585f463a32c739f2032e8e12d6c/lib.sh#L180 and
* https://github.com/nextcloud/vm/blob/a7fa9a2a11bf8585f463a32c739f2032e8e12d6c/lib.sh#L181

do the same thing

**Comment by enoch85** [comment]
> do the same thing

Yes, but we want to use existing functions if possible.

File: apps/fulltextsearch.sh
Comment (szaimen) [reply apps/fulltextsearch.sh:32]: Yes, that is what it is doing now. If any of the rules apply, it will start the reinstallation. Only if all are false, it will show the install screen  :)

**Comment by Ark74** [comment]
> Yes, but we want to use existing functions if possible.

well, create_certs is not done before.

**Comment by enoch85** [comment]
> well, create_certs is not done before.

It will be when we merge it. :)

Just tested the current version, and it installs fine, but I couldn't get it to find text inside documents (PDF).

**Comment by Ark74** [comment]
Interesting, maybe is just a db sync timing.

I'll setup both a new elastic and a opensearch later today to do some testing, it's amazingly late or amazingly early around here :sweat_smile:

File: lib.sh
Comment (Ark74) [inline lib.sh:182]: Maybe 
mv "$SCRIPTS"/opensearch_certs.sh "$OPNSDIR"

Comment (enoch85) [reply lib.sh:182]: I've arleady fixed it in your fork

File: lib.sh
Comment (enoch85) [reply lib.sh]: Let's merge first, then I can have a look with find/replace

File: lib.sh
Comment (Ark74) [reply lib.sh:182]: ok, seems like this could break things.
I'll be back in some hours. o/

File: apps/fulltextsearch.sh
Comment (enoch85) [inline apps/fulltextsearch.sh:16]: @szaimen aren't we using this in more places? :thinking: 

Maybe move to lib?

File: lib.sh
Comment (szaimen) [inline lib.sh:110]: are you sure it is a good idea to get this value every time we get the library?

Comment (enoch85) [reply lib.sh:110]: We might use it in more places, and I think it was a good variable.

Comment (szaimen) [reply lib.sh:110]: maybe wrap it in a function then?

Comment (enoch85) [reply lib.sh:110]: Sure, go ahead!

**Review by enoch85 — APPROVED** [review]

File: lib.sh
Comment (enoch85) [reply lib.sh:110]: I'll merge this now, we can fix details later.

**Comment by enoch85** [comment]
@szaimen If you want to fix the NCDOMAIN, please make a new PR, thanks! :pray:

**Comment by Ark74** [comment]
@enoch85 
> Just tested the current version, and it installs fine, but I couldn't get it to find text inside documents (PDF).

Seems like even when launching `fulltextsearch:index` it waits for a cron job to sync the db, the only way to force realtime results is using,

`occ fulltextsearch:live`

The only difference is that it will require to manually exit the application, as it remains open, while `fulltextsearch:index`, closes automatically once the index is complete and awaits the cron job for the update.

I think that leaving the default behavior is ok.

**Comment by enoch85** [comment]
Hmm, maybe something for a crontab, or service...

**Comment by Ark74** [comment]
I think it uses the nextcloud php cron job

**Pull request merged**
